### PR TITLE
[cs] Update C# grammar for additional C# 7 and C# 8 features

### PR DESCRIPTION
**Pull request by maikelsteneker**
## Describe the PR

This updates the Antlr grammar that is used for C# with the changes that were made in the [antlr/grammars-v4](https://github.com/antlr/grammars-v4) repository.

This includes:
* A fix for a case where empty commits were not tokenized correctly
* Support for digit separators `_` (C# 7.0)
* Support for binary literals (C# 7.0)
* Addition of the `unmanaged` keyword (C# 7.3)
* Addition of the null-coalescing assignment operator `??=` (C# 8.0)
* Addition of the range operator `..` (C# 8.0)
* Addition of the nullable preprocessor directive `nullable` (C# 8.0)

## Related issues

None that I'm aware of.

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by oowekyala — CHANGES_REQUESTED** [review]
Thanks for working on this! Please add a few tests for your changes though

**Comment by maikelsteneker** [comment]
Thanks for your comment! Unit tests have now been added.

File: pmd-cs/src/main/antlr4/net/sourceforge/pmd/lang/cs/antlr4/CSharpLexer.g4 [hunk 1]
```diff
@@ -19,7 +19,8 @@ private boolean verbatium;
 BYTE_ORDER_MARK: '\u00EF\u00BB\u00BF';
 
 SINGLE_LINE_DOC_COMMENT: '///' InputCharacter*    -> channel(COMMENTS_CHANNEL);
-DELIMITED_DOC_COMMENT:   '/**' .*? '*/'           -> channel(COMMENTS_CHANNEL);
+EMPTY_DELIMITED_DOC_COMMENT: '/***/'              -> channel(COMMENTS_CHANNEL);
```
Comment (oowekyala) [inline pmd-cs/src/main/antlr4/net/sourceforge/pmd/lang/cs/antlr4/CSharpLexer.g4:22]: @maikelsteneker Thanks for adding tests! The one last thing I have reservation is this change, as it's not obvious to me what is fixed here. Was `/***/` incorrectly tokenized previously? Maybe, add this to the test case and this should be gold.

Comment (maikelsteneker) [reply pmd-cs/src/main/antlr4/net/sourceforge/pmd/lang/cs/antlr4/CSharpLexer.g4:22]: I've added a test to cover this case. It fails with the old tokenizer, but passes with this updated one.

**Review by oowekyala — APPROVED** [review]
LGTM, thanks

**Pull request merged**